src/common/container_helpers.h

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) 2022, The Monero Project
+//
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without modification, are
+// permitted provided that the following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of
+//    conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list
+//    of conditions and the following disclaimer in the documentation and/or other
+//    materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+// Miscellaneous container helpers.
+
+#pragma once
+
+//local headers
+
+//third party headers
+
+//standard headers
+#include <algorithm>
+#include <functional>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+//forward declarations
+
+
+namespace tools
+{
+
+/// use operator< to get operator==
+/// WARNING: equality is not always implied by operator<, depending on implementation
+struct equals_from_less final
+{
+    template <typename T>
+    bool operator()(const T &a, const T &b) { return !(a < b) && !(b < a); }
+};
+/// note: uniqueness uses 'equals_from_less' to match the use of operator< when testing if sorted
+template <typename T>
+bool is_sorted_and_unique(const T& container)
+{
+    if (!std::is_sorted(container.begin(), container.end()))
+        return false;
+
+    if (std::adjacent_find(container.begin(), container.end(), equals_from_less{}) != container.end())
+        return false;
+
+    return true;
+}
```

Comment:
While this seems correct, silently defaulting to two < instead of a == will hinder performance (I suspect operator== is typically faster (or at least not slower) than even just one operator<). So this seems an odd choice. Is this because operator< is more likely to be available ?

Reply:
Checking if a container is sorted uses `operator<`. I want to use the same criteria for testing uniqueness. For example, I use this to check if tx output enotes are sorted by onetime address and also that all onetime addresses are unique.

Reply:
Tbh I am not convinced this is the best solution. Even though it is clean at the call-site, it is not clear what's happening. I may update this PR in the next couple days with an alternative.

Reply:
Ok I updated this to allow a user-defined comparison operator so you can add call-site clarity for non-trivial types by defining a comparison function.

Note that testing uniqueness with `adjacent_find` only works if you use the same comparison criteria as `is_sorted`.